rosbag2_py/src/rosbag2_py/_transport.cpp

```diff
@@ -239,6 +382,51 @@ class Player
   }
 
 protected:
+  bool wait_for_async(double timeout, const std::function<bool(double)> & wait_for_function)
+  {
+    if (!player_) {
+      throw std::runtime_error("Player is not initialized. Please use constructor with "
+        "storage and play options.");
+    }
+
+    const double wait_period =
+      (timeout < 0.0) ? SIGNAL_CHECK_INTERVAL : std::min(timeout, SIGNAL_CHECK_INTERVAL);
+
+    // We need to release the Global Interpreter Lock (GIL) to allow other Python threads to
+    // proceed. However, the GIL release doesn't result in Python checking for signals while
+    // we're in C++. We will wait for a short period of time and then check for signals in the loop.
+    // If any signals (Ctrl+C/SIGINT or SIGTERM) have been received, we will propagate them to
+    // Python.
+    bool finished = false;
+    while (!finished) {
+      {
+        // Release the Global Interpreter Lock (GIL) to allow other Python threads to proceed
+        py::gil_scoped_release release;
+        finished = wait_for_function(wait_period);
+      }
+      // The GIL release doesn't result in Python checking for signals while we're in C++, so we can
+      // do it here to simplify the Python layer: just periodically check for signals while waiting
+      // Wait for a short period of time
+      {
+        py::gil_scoped_acquire gil;
+        // If a signal (Ctrl+C/SIGINT or SIGTERM) has been received, propagate it to Python
+        if (PyErr_CheckSignals() != 0) {
+          throw py::error_already_set();
+        }
+      }
```

Comment:
Don't we already have the GIL here?

Reply:
Well, a better way to explicitly acquire it.

Reply:
ah okay, as long as it's fine if we already have the lock before this, then all good.